### PR TITLE
[WIP] add logabsdet(::UmfpackLU) method

### LINKED ISSUE #788: `logabsdet` unimplemented for sparse matrices
`logabsdet` is currently broken for matrices of type `SparseMatrixCSC`--or, more specifically, for their LU decomposition:

```julia
julia> using LinearAlgebra, SparseArrays

julia> A = sprandn(100, 100, 0.01);

julia> logabsdet(A)
ERROR: MethodError: no method matching logabsdet(::SuiteSparse.UMFPACK.UmfpackLU{Float64,Int64})
Closest candidates are:
  logabsdet(::SymTridiagonal; shift) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\tridiag.jl:441
  logabsdet(::UnitUpperTriangular{T,S} where S<:AbstractArray{T,2}) where T at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\triangular.jl:2652
  logabsdet(::UnitLowerTriangular{T,S} where S<:AbstractArray{T,2}) where T at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\triangular.jl:2653
  ...
Stacktrace:
 [1] logabsdet(::SparseMatrixCSC{Float64,Int64}) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\generic.jl:1573
 [2] top-level scope at none:1
```

### PR DESCRIPTION
**Pull request by ElOceanografo**
Will fix https://github.com/JuliaLang/LinearAlgebra.jl/issues/788.  This new method is a direct translation of the dense `logabsdet(::LU)` method in LinearAlgebra:https://github.com/JuliaLang/julia/blob/3365b4e903d61b628e2070e25c643cdfcf03c600/stdlib/LinearAlgebra/src/lu.jl#L457

It produces the correct answers, but is relatively slow and needs work on optimization.  The main bottleneck is extracting the factors via `umf_extract`, so directly indexing into the `UmfpackLU` struct should speed it up...though I'm not sure how to do that.

### CODE REVIEW

File: stdlib/SuiteSparse/src/umfpack.jl [hunk 1]
```diff
@@ -406,6 +406,29 @@ for itype in UmfpackIndexTypes
                         mx, mz, C_NULL, lu.numeric, umf_info)
             complex(mx[], mz[])
         end
+
+        function logabsdet(F::UmfpackLU{Float64, $itype})  # return log(abs(det)) and sign(det)
```
Comment (stevengj) [inline stdlib/SuiteSparse/src/umfpack.jl:410]: Why only `Float64`?

File: stdlib/SuiteSparse/test/umfpack.jl [hunk 2]
```diff
@@ -26,6 +26,7 @@ using LinearAlgebra: Adjoint, Transpose, SingularException
             @test (Diagonal(Rs) * A)[p,q] ≈ L * U
 
             det(lua) ≈ det(Array(A))
+            logabsdet(lua) ≈ logabsdet(Array(A))
```
Comment (stevengj) [inline stdlib/SuiteSparse/test/umfpack.jl:29]: Doesn't there need to be a `@test` in front of these?   Also, you need to check the sign separately, no?

**Comment by ViralBShah** [comment]
@ElOceanografo Maybe address the comments and get this merged?

**Comment by ElOceanografo** [comment]
Sorry, lost track of this! Comments were straightforward, but this is still very slow.  Benchmarking on my computer produces the following timings:
```julia
using LinearAlgebra, SparseArrays, BenchmarkTools
n = 2000
pnz = 0.01
A = sprandn(n, n, pnz); A = A'A + I
Adense = Matrix(A)
F = lu(A)

@btime logabsdet($A) # 549.719 ms (91 allocations: 278.57 MiB)
@btime logabsdet($F) # 79.753 ms (25 allocations: 86.26 MiB)
@btime logabsdet($Adense) # 46.192 ms (3 allocations: 30.53 MiB)
```
Profiling shows most of the time in the first line is spent [here](https://github.com/JuliaLang/julia/blob/3365b4e903d61b628e2070e25c643cdfcf03c600/stdlib/SuiteSparse/src/umfpack.jl#L327), while most of the time in the second line is spent extracting the fields from the LU factorization in [this line](https://github.com/JuliaLang/julia/blob/d8971760436f71e0c07f3a499236bc81fbcec3f8/stdlib/SuiteSparse/src/umfpack.jl#L413).  I know next to nothing about the internals of SuiteSparse, maybe someone who does can chime in?  Using the sparse factorization should in principle be faster than converting to a dense matrix and falling back on the default method, right?

**Comment by ViralBShah** [comment]
The problem with testing on random sparse matrices is that the LU factorization gets a lot of fill-in, making it almost dense. In this case, the filled up factors end up having 50% nonzeros. In reality, most sparse problems coming from a real application have more structure and significantly less fill-in.

**Comment by ViralBShah** [comment]
`UmfpackLU` is an opaque object, but this ends up spending a bunch of time converting the opaque object, when all you really want are the diagonals. I still think this should be ok to merge. Try some experiments with the identity matrix?

**Comment by ViralBShah** [comment]
Probably also rebase to master.

**Comment by ElOceanografo** [comment]
Running the same benchmark with `A = spdiagm(0 => ones(n))` is still slower than the dense version (189.999 μs (81 allocations: 1.11 MiB)).  If you know where the `UmfpackLU` object's layout is documented, I can try to figure out how to pick the diagonal values out directly.  But if you think this is worth merging as-is I can go ahead and rebase it.

**Comment by ViralBShah** [comment]
That will have to be in the UmfpackLU documentation or maybe even the Umfpack source code. I had trouble compiling this PR (following the gh command line) instructions and did end up rebasing it to master. Trying to get it working so I can try a few examples myself.

**Comment by ElOceanografo** [comment]
I did some digging in the Umfpack documentation, and it should be possible to extract just the `U` matrix and scaling factors `Rs` by passing null values in for the other arguments (see section 13, top of page 94, in the pdf here: https://fossies.org/linux/SuiteSparse/UMFPACK/Doc/UMFPACK_UserGuide.pdf).  That isn't currently possible through the `SuiteSparse.UMFPACK` wrappers, which is probably worth a separate issue.

**Comment by ViralBShah** [comment]
Something has gone wrong with this PR.

**Comment by ElOceanografo** [comment]
I tried to rebase it to master last night and may have messed it up.  I can try to fix it, or if it's easier I could close it and open a new one with the same changes...

**Comment by ViralBShah** [comment]
If its easier to do a new PR, just go for it.

**Comment by ElOceanografo** [comment]
Ok, closing this, will open a new one in a second.

**Pull request closed without merging**